security-jwt/src/main/java/io/micronaut/security/token/jwt/validator/JwtValidator.java

```diff
@@ -191,38 +193,87 @@ private boolean hasAtLeastTwoDots(String token) {
+        sortedConfigs.sort(comparator(algorithm, jwt.getHeader().getKeyID()));
 
-        for (SignatureConfiguration config: sortedConfigs) {
-            try {
-                boolean verified = config.verify(jwt);
-                if (!verified) {
-                    if (LOG.isDebugEnabled()) {
-                        LOG.debug("JWT Signature verification failed: {}", jwt.getParsedString());
-                    }
-                } else {
-                    return Optional.of(jwt);
-                }
-            } catch (final JOSEException e) {
-                if (LOG.isDebugEnabled()) {
-                    LOG.debug("Verification failed with signature configuration: {}, passing to the next one", config);
+        Optional<JWT> optionalJWT = validate(jwt, sortedConfigs);
+        if (optionalJWT.isPresent()) {
+            return optionalJWT;
+        }
+
+        // If any of the signature configurations is a JwksCache, evict the cache and attempt to verify again
+        for (SignatureConfiguration c : sortedConfigs.stream()
+                .filter(signConf -> signConf instanceof JwksCache)
+                .collect(Collectors.toList())) {
+            if (((JwksCache) c).isJwksCacheExpired().orElse(false)) {
+                ((JwksCache) c).clearJwksCache();
+                optionalJWT = validate(jwt, c);
+                if (optionalJWT.isPresent()) {
+                    return optionalJWT;
                 }
             }
         }
 
         if (LOG.isDebugEnabled() && signatures.isEmpty()) {
             LOG.debug("JWT is signed and no signature configurations -> not verified");
         }
+        return Optional.empty();
+    }
+
+    private Optional<JWT> validate(SignedJWT jwt, SignatureConfiguration signatureConfiguration) {
+        try {
+            boolean verified = signatureConfiguration.verify(jwt);
+            if (!verified) {
+                if (LOG.isDebugEnabled()) {
+                    LOG.debug("JWT Signature verification failed: {}", jwt.getParsedString());
+                }
+            } else {
+                return Optional.of(jwt);
+            }
+        } catch (final JOSEException e) {
+            if (LOG.isDebugEnabled()) {
+                LOG.debug("Verification failed with signature configuration: {}, passing to the next one", signatureConfiguration);
+            }
+        }
+        return Optional.empty();
+    }
 
+    private Optional<JWT> validate(SignedJWT jwt, List<SignatureConfiguration> signatureConfigurations) {
+        for (SignatureConfiguration config: signatureConfigurations) {
+            Optional<JWT> optionalJWT = validate(jwt, config);
+            if (optionalJWT.isPresent()) {
+                return optionalJWT;
+            }
+        }
         return Optional.empty();
     }
 
-    private static Comparator<SignatureConfiguration> comparator(JWSAlgorithm algorithm) {
+    private static int compareMatchingKids(SignatureConfiguration sig,
```

Comment:
Is this supposed to mean `Key IDs`? If so I would have that has the name because what this method does isn't clear by the name